ui/library/src/context/PageRenderContext.ts

```diff
@@ -0,0 +1,243 @@
+import * as React from 'react';
+import { pdfjs } from 'react-pdf';
+
+import { Nullable } from '../components/types/utils';
+import { logProviderWarning } from '../utils/provider';
+import { PageRotation } from '../utils/rotate';
+
+export type RenderState = {
+  promise: Promise<string>;
+  objectURL: Nullable<string>;
+};
+
+export type PageNumberToRenderStateMap = Map<number, RenderState>;
+
+export interface IPageRenderContext {
+  pageRenderStates: PageNumberToRenderStateMap;
+  getObjectURLForPage: (args: { pageNumber?: number; pageIndex?: number }) => Nullable<string>;
+  isBuildingObjectURLForPage: (args: { pageNumber?: number; pageIndex?: number }) => boolean;
+  buildObjectURLForPage: (args: { pageNumber?: number; pageIndex?: number }) => Promise<string>;
+}
+
+export const PageRenderContext = React.createContext<IPageRenderContext>({
+  pageRenderStates: new Map(),
+  getObjectURLForPage: args => {
+    logProviderWarning(`getObjectURLForPage(${JSON.stringify(args)})`, 'PageRenderContext');
+    return null;
+  },
+  isBuildingObjectURLForPage: args => {
+    logProviderWarning(`isBuildingObjectURLForPage(${JSON.stringify(args)})`, 'PageRenderContext');
+    return false;
+  },
+  buildObjectURLForPage: args => {
+    logProviderWarning(`buildObjectURLForPage(${JSON.stringify(args)})`, 'PageRenderContext');
+    return Promise.resolve('');
+  },
+});
+
+export function usePageRenderContextProps({
+  pdfDocProxy,
+  scale,
+  rotation,
+  zoomMultiplier,
+  visiblePageRatios,
+}: {
+  pdfDocProxy?: pdfjs.PDFDocumentProxy;
+  scale: number;
+  rotation: PageRotation;
+  zoomMultiplier: number;
+  visiblePageRatios: Map<number, number>;
+}): IPageRenderContext {
+  const [pageRenderStates, _setPageRenderStates] = React.useState<PageNumberToRenderStateMap>(
+    () => {
+      const map = new Map();
+      Object.freeze(map);
+      return map;
+    }
+  );
+
+  // Because rendering a page is async, we will lose the current pageRenderStates
+  // This ref trick allows the latest to be accessible when the objectURL is ready
+  const pageRenderStatesRef = React.useRef(pageRenderStates);
+  const setPageRenderStates = React.useCallback(
+    (pageRenderStates: PageNumberToRenderStateMap) => {
+      pageRenderStatesRef.current = pageRenderStates;
+      _setPageRenderStates(pageRenderStates);
+    },
+    [pageRenderStatesRef]
+  );
+
+  const isBuildingObjectURLForPage = React.useCallback(
+    ({ pageNumber, pageIndex }: { pageNumber?: number; pageIndex?: number }): boolean => {
+      if (typeof pageIndex === 'number') {
+        pageNumber = pageIndex + 1;
+      }
+      if (typeof pageNumber !== 'number') {
+        return false;
+      }
```

Comment:
What are the chances for pageIndex/pageNumber not being 'number' type?

Reply:
@yensung there is a case it can be undefined so thats why there is this check

Reply:
If so, why don't you check if pageNumber is presented or not? The cause for pageNumber being undefined is when it is not presented. But when it is presented, pageNumber is guaranteed to be number, right?

Reply:
@yensung i have do a check for if its existing or not but it ended up not generating the url at all.